cosmwasm/packages/wasmi-runtime/src/wasm/module_cache.rs

```diff
@@ -0,0 +1,121 @@
+use std::collections::HashMap;
+use std::sync::SgxRwLock;
+
+use lazy_static::lazy_static;
+use log::*;
+use parity_wasm::elements;
+use parity_wasm::elements::Module;
+use wasmi::{ModuleInstance, ModuleRef};
+
+use enclave_ffi_types::EnclaveError;
+
+use crate::crypto::HASH_SIZE;
+use crate::wasm::types::ContractCode;
+
+use super::gas::{gas_rules, WasmCosts};
+use super::memory::validate_memory;
+use super::runtime::{create_builder, WasmiImportResolver};
+
+lazy_static! {
+    static ref MODULE_CACHE: SgxRwLock<HashMap<[u8; HASH_SIZE], wasmi::Module>> =
+        SgxRwLock::new(HashMap::new());
+}
+
+pub fn create_module_instance(contract_code: ContractCode) -> Result<ModuleRef, EnclaveError> {
+    let code_hash = contract_code.hash();
+    match get_module_instance(&code_hash) {
+        Some(Ok(module_ref)) => return Ok(module_ref),
+        None => {} // continue
+
+        // If the stored module failed to process for some reason, remove it.
+        // Shouldn't happen because we already compiled it before.
+        Some(Err(_)) => {
+            MODULE_CACHE.write().unwrap().remove(&code_hash);
+        }
+    }
+
+    let mut cache = MODULE_CACHE.write().unwrap();
+    match cache.get(&code_hash).map(create_instance) {
+        Some(Ok(module_ref)) => return Ok(module_ref),
+        None => {} // continue
+
+        // If the stored module failed to process for some reason, remove it.
+        // Shouldn't happen because we already compiled it before.
+        Some(Err(_)) => {
+            cache.remove(&code_hash);
+        }
+    }
```

Comment:
Why do you need this code section here again?

Reply:
This code section, as well as the usage or RwLock which you asked about in the other comment, are made with enclave multithreading in mind. So, the RwLock lets you have many concurrent readers, but only one writer. And in this code section, if we were waiting for the write lock, then we check if another thread already wrote to the same location while we were waiting.

check -> lock -> check -> process -> write

Reply:
Yep that's what I thought. Cool :)